In [21]:
# Đọc dữ liệu từ tệp CSV
data = pd.read_csv('D:\\Python\\MachineLearning\\MachineLearning\\week06\\data\\Analysis-Portfolio-Task-Data.csv')

# Chuyển đổi cột 'Gender' thành 0 và 1
data['Gender'] = data['Gender'].replace({'Male': 0, 'Female': 1})

# Chuyển đổi cột 'Store Type' thành 0, 1 và 2
store_type_mapping = {'Convenient Stores': 0, 'Superstore': 1, 'Online': 2}
data['Store Type'] = data['Store Type'].replace(store_type_mapping)

# In dữ liệu đã được chuyển đổi
print(data.head())
X = data.drop(columns=['CustomerID', 'Shopping Basket'])
y = data['Shopping Basket']
# Phân chia dữ liệu thành tập huấn luyện và tập validation
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=15, random_state=42)

# Huấn luyện mô hình hồi quy tuyến tính
model = LinearRegression()
model.fit(X_train, y_train)

# Dự đoán trên tập validation
y_pred = model.predict(X_val)

# Đánh giá mô hình
mse = mean_squared_error(y_val, y_pred)
r_squared = model.score(X_val, y_val)

print("Mean Squared Error:", mse)
print("R-squared:", r_squared)
print("Coefficients:", model.coef_)


   CustomerID  Shopping Basket  Gender  Age  Store Type  Value Products  \
0    20358063            48.81       0   26           0               8   
1    24635139            33.44       1   33           1               6   
2    27584479           131.57       0   56           2              35   
3    28008212            20.02       0   27           0               0   
4    29130973            95.54       1   55           2              38   

   Brand Products  Top Fresco Products  
0               2                    1  
1               5                    1  
2               8                   12  
3               1                    1  
4              18                   20  
Mean Squared Error: 476.2714169776164
R-squared: 0.6952340998457984
Coefficients: [ 5.01345458  0.08483512 11.53477706  0.94619626  0.78677569  2.77054881]


Mean Squared Error (MSE) có giá trị là 476.27. MSE đo lường sự chênh lệch giữa các giá trị thực tế và các giá trị được dự đoán bởi mô hình, với giá trị càng thấp càng tốt. Trong trường hợp này, MSE có giá trị lớn, điều này cho thấy mô hình có sự sai lệch lớn giữa dự đoán và giá trị thực tế.

R-squared có giá trị là 0.695, đây là một phép đo đánh giá mức độ mô hình giải thích dữ liệu. R-squared càng gần 1 thì mô hình càng tốt. Trong trường hợp này, giá trị R-squared là 0.695, tức là mô hình chỉ giải thích được khoảng 69.5% sự biến thiên của dữ liệu, điều này cho thấy mô hình không phản ánh được một phần lớn sự biến thiên của dữ liệu.

Hệ số (coefficients) của mô hình thể hiện mối quan hệ giữa các biến đầu vào và biến mục tiêu. Trong trường hợp này, các hệ số cho biết độ ảnh hưởng của mỗi biến đầu vào đối với biến mục tiêu. Đối với mỗi biến, hệ số càng lớn thì ảnh hưởng của biến đó đối với biến mục tiêu càng cao. Điều này có thể giúp định rõ những yếu tố quan trọng nhất đối với mục tiêu của mô hình.

Tóm lại, mặc dù mô hình có khả năng giải thích được một phần nào đó của dữ liệu (R-squared không quá thấp), nhưng MSE cao cho thấy mô hình vẫn cần được cải thiện để dự đoán chính xác hơn. Cần xem xét các biện pháp cải thiện mô hình như thêm biến, điều chỉnh siêu tham số của mô hình hoặc sử dụng các phương pháp hồi quy khác.

In [22]:
    import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Perceptron, LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score

# Đọc dữ liệu từ tệp CSV
data = pd.read_csv('D:\\Python\\MachineLearning\\MachineLearning\\week06\\data\\Analysis-Portfolio-Task-Data.csv')

# Chuyển đổi cột 'Gender' thành 0 và 1
data['Gender'] = data['Gender'].replace({'Male': 0, 'Female': 1})

# Chuyển đổi cột 'Store Type' thành 0, 1 và 2
store_type_mapping = {'Convenient Stores': 0, 'Superstore': 1, 'Online': 2}
data['Store Type'] = data['Store Type'].replace(store_type_mapping)

# Tạo cột nhãn
data['Label'] = data['Shopping Basket'].apply(lambda x: 1 if x > 50 else -1)

# In dữ liệu đã được chuyển đổi
print(data.head())

X = data.drop(columns=['CustomerID', 'Shopping Basket', 'Label'])
y = data['Label']

# Phân chia dữ liệu thành tập huấn luyện và tập validation
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

# a. Xây dựng mô hình Perceptron
perceptron_model = Perceptron()
perceptron_model.fit(X_train, y_train)
perceptron_y_pred = perceptron_model.predict(X_val)
perceptron_accuracy = accuracy_score(y_val, perceptron_y_pred)
print("Perceptron Accuracy:", perceptron_accuracy)

# b. Xây dựng mô hình Logistic Regression
logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)
logistic_y_pred = logistic_model.predict(X_val)
logistic_accuracy = accuracy_score(y_val, logistic_y_pred)
print("Logistic Regression Accuracy:", logistic_accuracy)

# c. Xây dựng mô hình Naive Bayes
naive_bayes_model = GaussianNB()
naive_bayes_model.fit(X_train, y_train)
naive_bayes_y_pred = naive_bayes_model.predict(X_val)
naive_bayes_accuracy = accuracy_score(y_val, naive_bayes_y_pred)
print("Naive Bayes Accuracy:", naive_bayes_accuracy)


   CustomerID  Shopping Basket  Gender  Age  Store Type  Value Products  \
0    20358063            48.81       0   26           0               8   
1    24635139            33.44       1   33           1               6   
2    27584479           131.57       0   56           2              35   
3    28008212            20.02       0   27           0               0   
4    29130973            95.54       1   55           2              38   

   Brand Products  Top Fresco Products  Label  
0               2                    1     -1  
1               5                    1     -1  
2               8                   12      1  
3               1                    1     -1  
4              18                   20      1  
Perceptron Accuracy: 0.8
Logistic Regression Accuracy: 1.0
Naive Bayes Accuracy: 0.9333333333333333


Độ chính xác (Accuracy):

Logistic Regression có độ chính xác cao nhất trên tập validation, với tất cả các điểm dữ liệu được phân loại đúng. Điều này cho thấy mô hình Logistic Regression có khả năng phân loại tốt nhất trong trường hợp này.
Naive Bayes đứng thứ hai với độ chính xác lớn, tuy nhiên, thấp hơn so với Logistic Regression.
Perceptron đạt được độ chính xác tương đối cao nhưng thấp hơn so với hai mô hình còn lại.
Phù hợp với dữ liệu:

Logistic Regression và Naive Bayes thường hoạt động tốt trên dữ liệu có cấu trúc tuyến tính hoặc dữ liệu mà các biến độc lập có ảnh hưởng tuyến tính đến biến mục tiêu.
Perceptron cũng là một mô hình tuyến tính, nhưng nó có thể không hoạt động tốt trên các tập dữ liệu phức tạp hơn hoặc dữ liệu không tuyến tính.
Khả năng giải thích:

Logistic Regression cho phép xác định rõ ràng mức độ ảnh hưởng của từng biến độc lập đến biến phụ thuộc thông qua hệ số của chúng. Điều này giúp hiểu rõ hơn về cách mà mỗi biến ảnh hưởng đến dự đoán của mô hình.
Naive Bayes giả định độc lập giữa các biến độc lập, do đó các hệ số của mô hình không cung cấp thông tin về mức độ ảnh hưởng của mỗi biến đến dự đoán. Tuy nhiên, mô hình này có thể hoạt động tốt trong nhiều trường hợp thực tế và dễ dàng hiểu được cách hoạt động của nó.
Perceptron cũng cung cấp các hệ số, nhưng chúng không cho biết mức độ ảnh hưởng của từng biến một cách rõ ràng như trong Logistic Regression.
Dựa trên các yếu tố trên, trong trường hợp này, Logistic Regression là lựa chọn tốt nhất do có độ chính xác cao nhất và khả năng giải thích tốt nhất về mối quan hệ giữa các biến độc lập và biến phụ thuộc.








Dựa trên kết quả của các mô hình phân loại:

Mô hình Perceptron có độ chính xác là 0.8, tức là khoảng 80% các điểm dữ liệu trong tập validation được phân loại đúng. Mô hình này hoạt động khá tốt, nhưng vẫn cần cải thiện để đạt được độ chính xác cao hơn.

Mô hình Logistic Regression có độ chính xác là 1.0, tức là tất cả các điểm dữ liệu trong tập validation được phân loại đúng. Đây là một kết quả rất tốt và cho thấy mô hình này hoạt động rất hiệu quả trên dữ liệu đã cho.

Mô hình Naive Bayes có độ chính xác là 0.933, tức là khoảng 93.3% các điểm dữ liệu trong tập validation được phân loại đúng. Mô hình này cũng hoạt động tốt, nhưng có thể cần một số điều chỉnh để cải thiện độ chính xác.

So sánh giữa các mô hình:

Mô hình Logistic Regression đạt được độ chính xác cao nhất trên tập validation, với tất cả các điểm dữ liệu được phân loại đúng. Điều này cho thấy rằng mô hình Logistic Regression phù hợp nhất với dữ liệu đã cho trong trường hợp này.

Mô hình Naive Bayes cũng cho kết quả tốt với độ chính xác cao, nhưng thấp hơn so với Logistic Regression.

Mô hình Perceptron cũng hoạt động tốt, nhưng cần cải thiện độ chính xác để đạt được kết quả tốt nhất.

Tóm lại, mô hình Logistic Regression là lựa chọn tốt nhất trong trường hợp này với độ chính xác cao nhất trên tập validation.
